src/debugger/crash_screen.c

```diff
@@ -0,0 +1,555 @@
+    80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95,
+};
+
+u32 gCrashScreenFont[GLYPH_HEIGHT * FONT_ROWS * 2 + 1] = {
+    #include "textures/crash_custom/crash_screen_font.ia1.inc.c"
+};
+
+u8 crashPage = 0;
+u8 updateBuffer = TRUE;
+
+
+char *gCauseDesc[18] = {
+    "Interrupt",
+    "TLB modification",
+    "TLB exception on load",
+    "TLB exception on store",
+    "Address error on load",
+    "Address error on store",
+    "Bus error on inst.",
+    "Bus error on data",
+    "Failed Assert: See Assert Page",
+    "Breakpoint exception",
+    "Reserved instruction",
+    "Coprocessor unusable",
+    "Arithmetic overflow",
+    "Trap exception",
+    "Virtual coherency on inst.",
+    "Floating point exception",
+    "Watchpoint exception",
+    "Virtual coherency on data",
+};
+
+char *gFpcsrDesc[6] = {
+    "Unimplemented operation", "Invalid operation", "Division by zero", "Overflow", "Underflow",
+    "Inexact operation",
+};
+
+
+
+extern u64 osClockRate;
+
+struct {
+    OSThread thread;
+    u64 stack[THREAD2_STACK / sizeof(u64)];
+    OSMesgQueue mesgQueue;
+    OSMesg mesg;
+    u16 *framebuffer;
+    u16 width;
+    u16 height;
+} gCrashScreen;
+
+static u16 gCrashScreenColor = 0xFFFF;
+
+static void set_text_color(u32 r, u32 g, u32 b) {
+    gCrashScreenColor = GPACK_RGBA5551(r, g, b, 255);
+}
+static void reset_text_color(void) {
+    gCrashScreenColor = 0xFFFF;
+}
+
+void crash_screen_draw_rect(s32 x, s32 y, s32 w, s32 h) {
+    u16 *ptr;
+    s32 i, j;
+
+    ptr = gCrashScreen.framebuffer + gCrashScreen.width * y + x;
+    for (i = 0; i < h; i++) {
+        for (j = 0; j < w; j++) {
+            *ptr = 0x0001;
+            ptr++;
+        }
+        ptr += gCrashScreen.width - w;
+    }
+}
+
+void crash_screen_draw_glyph(s32 x, s32 y, s32 glyph) {
+    const u32 *data;
+    u16 *ptr;
+    u32 bit;
+    u32 rowMask;
+    s32 i, j;
+
+    if (glyph > 0x7F) return;
+
+    data = &gCrashScreenFont[((glyph&0xF)*GLYPH_HEIGHT * 2) + (glyph >= 64)];
+
+    ptr = gCrashScreen.framebuffer + gCrashScreen.width * y + x;
+
+    u16 color = gCrashScreenColor;
+
+    for (i = 0; i < GLYPH_HEIGHT; i++) {
+        bit = 0x80000000U >> ((glyph >> 4) * GLYPH_WIDTH);
+        rowMask = *data++;
+        data ++;
+
+        for (j = 0; j < (GLYPH_WIDTH); j++) {
+            if (bit & rowMask) {
+                *ptr = color;
+            }
+            ptr++;
+            bit >>= 1;
+        }
+        ptr += gCrashScreen.width - (GLYPH_WIDTH);
+    }
+}
+
+static char *write_to_buf(char *buffer, const char *data, size_t size) {
+    return (char *) memcpy(buffer, data, size) + size;
+}
+
+void crash_screen_print(s32 x, s32 y, const char *fmt, ...) {
+    char *ptr;
+    u32 glyph;
+    s32 size;
+    s32 xOffset = x;
+    char buf[0x108];
+    bzero(&buf, sizeof(buf));
+
+    va_list args;
+    va_start(args, fmt);
+
+    size = _Printf(write_to_buf, buf, fmt, args);
```

Comment:
Is `vsprintf` not available? Perhaps we should borrow Falco's libc implementation. It is pretty good although needs some fluff stripped.